File: solutions/problem_1298_maximum_candies_you_can_get_from_boxes/solution.py
"""
Problem 1298: Maximum Candies You Can Get from Boxes
==================================================
Difficulty: Hard
Tags: Array, Breadth-First Search, Graph Theory
Pattern: BFS Graph Traversal

Time Complexity:  O(V + E)
Space Complexity: O(V)
"""

class Solution:
    def maxCandies(self, status: List[int], candies: List[int], keys: List[List[int]], containedBoxes: List[List[int]], initialBoxes: List[int]) -> int:
        # BFS on graph - O(V+E) time
        from collections import deque
        if not status:
            return 0
        visited = set()
        queue = deque([0])
        visited.add(0)
        dist = 0
        while queue:
            for _ in range(len(queue)):
                node = queue.popleft()
                # Process node
            dist += 1
        return dist

File: solutions/problem_1298_maximum_candies_you_can_get_from_boxes/solution.cpp
/*
 * Problem 1298: Maximum Candies You Can Get from Boxes
 * ==================================================
 * Difficulty: Hard
 * Tags: Array, Breadth-First Search, Graph Theory
 * Pattern: BFS Graph Traversal
 *
 * Time Complexity:  O(V + E)
 * Space Complexity: O(V)
 */

#include <queue>
#include <string>
#include <unordered_set>
#include <vector>
using namespace std;

class Solution {
public:
    int maxCandies(vector<int>& status, vector<int>& candies, vector<vector<int>>& keys, vector<vector<int>>& containedBoxes, vector<int>& initialBoxes) {
        // BFS on graph - O(V+E) time
        if (status.empty()) return 0;
        queue<int> q;
        unordered_set<int> visited;
        q.push(0);
        visited.insert(0);
        int dist = 0;
        while (!q.empty()) {
            int sz = q.size();
            for (int i = 0; i < sz; i++) {
                int node = q.front(); q.pop();
                // Process node
            }
            dist++;
        }
        return dist;
    }
};
